src/coreclr/jit/llvm.cpp

```diff
@@ -175,6 +213,70 @@ Value* buildAdd(llvm::IRBuilder<>& builder, GenTree* node, Value* op1, Value* op
     return mapTreeIdValue(node->gtTreeID, builder.CreateGEP(op1, op2));
 }
 
+FunctionType* getFunctionSignature(CORINFO_CONST_LOOKUP callTarget)
+{
+    return FunctionType::get(Type::getInt8PtrTy(_llvmContext), ArrayRef<Type*>(Type::getInt8PtrTy(_llvmContext)), false);
+}
+
+Value* genTreeAsLlvmType(GenTree* gt, Type* t)
+{
+    Value* v = getTreeIdValue(gt);
+    if (v->getType() == t)
+        return v;
+
+    if (gt->OperGet() == GT_CNS_INT && gt->gtType == var_types::TYP_INT)
+    {
+        return _builder->getInt({(unsigned int)t->getPrimitiveSizeInBits().getFixedSize(), (uint64_t)gt->AsIntCon()->gtIconVal, true});
+    }
+    failFunctionCompilation();
+}
+
+llvm::Value* buildUserFuncCall(GenTreeCall* gtCall, llvm::IRBuilder<>& builder)
+{
+    const char* symbolName = (*_getMangledSymbolName)(_thisPtr, gtCall->gtEntryPoint.handle);
+    // TODO: how to detect RuntimeImport ?
+    if (!strcmp(symbolName, "S_P_CoreLib_System_Runtime_RuntimeImports__MemoryBarrier") ||
+        !strcmp(symbolName, "S_P_CoreLib_System_Runtime_RuntimeImports__RhCollect"))
+    {
+        failFunctionCompilation();
+    }
+
+    (*_addCodeReloc)(_thisPtr, gtCall->gtEntryPoint.handle);
+    Function* llvmFunc = _module->getFunction(symbolName);
+    if (llvmFunc == nullptr)
+    {
+        CORINFO_SIG_INFO sigInfo;
+        _compiler->eeGetMethodSig(gtCall->gtCallMethHnd, &sigInfo);
+        CORINFO_ARG_LIST_HANDLE sigArgs = sigInfo.args;
+
+        // assume ExternalLinkage, if the function is defined in the clrjit module, then it is replaced and an extern
+        // added to the Ilc module
+        llvmFunc = Function::Create(getFunctionTypeForMethodHandle(gtCall->gtCallMethHnd), Function::ExternalLinkage,
+                                    0U, symbolName, _module);
+    }
+    std::vector<llvm::Value*> argVec;
+
+    // shadowstack arg first
+    argVec.push_back(_function->getArg(0));
+
+    // TODO Is Args() more appropriate than Operands() ?
```

Comment:
I am not sure. @AndyAyersMS What is the difference between gtCall `Args()` vs `Operands()`?

Reply:
`Operands()` is a superset of `Args()`, it gives you (in order)
* `this` pointer, if any;
* `Args()`
* `LateArgs()`
* call cookie (for CT_INDIRECT calls)
* call address (for CT_INDIRECT calls)
* control expression (for other sorts of indirect calls)
 
See 
https://github.com/dotnet/runtime/blob/ab820d7a68b777a2da3dd1d89d4018e9e3d3a0fe/src/coreclr/jit/compiler.hpp#L4432-L4472